File: tree/449. Serialize and Deserialize BST/js/main.js
//  BST could be constructed from preorder or postorder traversal only.

// Serialization could be easily implemented with both strategies
// But for optimal deserialization better to choose the postorder traversal because member/global/static variables are not allowed here.

// Time O(N)
// Space O(N)

/**
 * Encodes a tree to a single string.
 *
 * @param {TreeNode} root
 * @return {string}
 */
const serialize = root => {
  if (root == null) return [];

  let res = [];

  dfs(root);

  return res;

  function dfs(node) {
    if (node == null) {
      return;
    }

    dfs(node.left);
    dfs(node.right);
    res.push(node.val);
  }
};

/**
 * Decodes your encoded data to tree.
 *
 * @param {Array} data
 * @return {TreeNode}
 */
const deserialize = data => {
  if (data.length == 0) return null;

  let INF = Number.MAX_VALUE;

  return dfs(-INF, INF);

  // Post-order Traversal
  // left-right-root
  function dfs(lower, upper) {
    if (data.length == 0 || data[data.length - 1] < lower || data[data.length - 1] > upper) {
      return null;
    }

    let val = data.pop();
    let node = new TreeNode(val);

    node.right = dfs(val, upper);
    node.left = dfs(lower, val);

    return node;
  }
};

File: tree/449. Serialize and Deserialize BST/py/main.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Codec:

    def serialize(self, root: TreeNode) -> str:
        """Encodes a tree to a single string.
        """
        if root == None:
            return ''

        res = []

        def dfs(node):
            if node == None:
                return
            dfs(node.left)
            dfs(node.right)
            res.append(node.val)
        
        dfs(root)

        return ' '.join(map(str, res))
        


    def deserialize(self, data: str) -> TreeNode:
        """Decodes your encoded data to tree.
        """
        if not data:
            return None
        
        def dfs(lower = -float('inf'), upper = float('inf')):
            if not data or data[-1] < lower or data[-1] > upper:
                return None

            val = data.pop()
            node = TreeNode(val)

            node.right = dfs(val, upper)
            node.left = dfs(lower, val)

            return node 
        
        data = [int(x) for x in data.split(' ') if x]
        
        return dfs(0, len(data))
